### ISSUE TITLE
[JENKINS-44339] Invalid syntax error:  Missing required parameter: "image"

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/pipeline-model-definition-plugin#1077]
<p>When I use a docker agent in a parallel branch, I get a syntax error saying the image parameter is missing even though it is clearly present.</p>

<p>I have the following pipeline script:</p>

<pre>
pipeline {
  agent any
  stages {
    stage(<span class="code-quote">'Test Ubuntu'</span>) {
      steps {
parallel(
  <span class="code-quote">"Test Ubuntu"</span>: {
    agent {
      docker {
image <span class="code-quote">'jess/zsh'</span>
      }
    }
    sh <span class="code-quote">'zsh allthethings.sh'</span>
    
  },
  <span class="code-quote">"Test CentOS"</span>: {
    agent {
      docker {
image <span class="code-quote">'jess/zsh'</span>
      }
      }
    sh <span class="code-quote">'zsh allthethings.sh'</span>
    
  }
)
      }
    }
    stage(<span class="code-quote">'notify'</span>) {
      steps {
echo <span class="code-quote">'Notifying the folks!'</span>
      }
    }
  }
</pre>

<p>But I get the following error:</p>

<pre>
org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
WorkflowScript: 9: Missing required parameter: <span class="code-quote">"image"</span> @ line 9, column 15.
 docker {
 ^

WorkflowScript: 18: Missing required parameter: <span class="code-quote">"image"</span> @ line 18, column 15.
 docker {
 ^

2 errors

	at org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:310)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1073)
	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:591)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:569)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:546)
	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:298)
	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:268)
	at groovy.lang.GroovyShell.parseClass(GroovyShell.java:688)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:700)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.reparse(CpsGroovyShell.java:116)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.parseScript(CpsFlowExecution.java:430)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.start(CpsFlowExecution.java:393)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:240)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:405)
Finished: FAILURE
</pre>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/recampbell.png" title="recampbell's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=recampbell">recampbell</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44339" target="_blank">Invalid syntax error:  Missing required parameter: "image"</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer">abayer</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: pipeline-model-definition-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2017-05-18T13:03:10+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 1
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>When I use a docker agent in a parallel branch, I get a syntax error saying the image parameter is missing even though it is clearly present.</p>

<p>I have the following pipeline script:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">pipeline {
  agent any
  stages {
    stage(<span class="code-quote">'Test Ubuntu'</span>) {
      steps {
        parallel(
          <span class="code-quote">"Test Ubuntu"</span>: {
            agent {
              docker {
                image <span class="code-quote">'jess/zsh'</span>
              }
            }
            sh <span class="code-quote">'zsh allthethings.sh'</span>
            
          },
          <span class="code-quote">"Test CentOS"</span>: {
            agent {
              docker {
                image <span class="code-quote">'jess/zsh'</span>
              }
              }
            sh <span class="code-quote">'zsh allthethings.sh'</span>
            
          }
        )
      }
    }
    stage(<span class="code-quote">'notify'</span>) {
      steps {
        echo <span class="code-quote">'Notifying the folks!'</span>
      }
    }
  }
</pre>
</div></div>

<p>But I get the following error:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
WorkflowScript: 9: Missing required parameter: <span class="code-quote">"image"</span> @ line 9, column 15.
                 docker {
                 ^

WorkflowScript: 18: Missing required parameter: <span class="code-quote">"image"</span> @ line 18, column 15.
                 docker {
                 ^

2 errors

	at org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:310)
	at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1073)
	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:591)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:569)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:546)
	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:298)
	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:268)
	at groovy.lang.GroovyShell.parseClass(GroovyShell.java:688)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:700)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.reparse(CpsGroovyShell.java:116)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.parseScript(CpsFlowExecution.java:430)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.start(CpsFlowExecution.java:393)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:240)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:405)
Finished: FAILURE
</pre>
</div></div></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>pipeline-model-definition 1.1.4</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/recampbell.png" title="recampbell's avatar" /> recampbell:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/44339?focusedId=300072&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-300072">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44339?focusedId=300072&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-300072">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Apparently agents in parallel aren't supported, so doesn't make sense to fix the error message because we are just going to do this: <a href="https://issue-redirect.jenkins.io/issue/41334" title="Support parallel execution of stages in Declarative" class="issue-link" data-issue-key="JENKINS-41334"><del>JENKINS-41334</del></a></p>
</pre>
</details>

</ul>
</details>

<p>Apparently agents in parallel aren't supported, so doesn't make sense to fix the error message because we are just going to do this: <a href="https://issue-redirect.jenkins.io/issue/41334" title="Support parallel execution of stages in Declarative" class="issue-link" data-issue-key="JENKINS-41334"><del>JENKINS-41334</del></a></p>
